File: python/fib.py
#!/usr/bin/python3
import timeit

fibval = int(input("Enter a vlue for the fibonacci sequence: "))
memo_list = [None] * (fibval + 1)
memo_list2 = [None] * (fibval + 1)
temp = 0

def fib(x):
    if(x < 0): return 0
    if(x <= 1): return x
    return fib(x-1) + fib(x-2)

def fib2(x):
    if(x < 0): return 0
    if(x == 0 or x == 1): return x
    return fib2(x-1) + fib2(x-2)

def fib_memo(x):
    if(x < 0): return 0
    if(x <= 1): 
        memo_list[x] = x
        return x
    if(memo_list[x] != None): 
        return memo_list[x]
    ret = fib_memo(x-1) + fib_memo(x-2)
    memo_list[x] = ret
    return ret

def fib2_memo(x):
    if(x < 0): 
        return 0
    if(x == 0 or x == 1): 
        memo_list2[x] = x
        return x
    if(memo_list2[x] != None): 
        return memo_list2[x]
    ret = fib2_memo(x-1) + fib2_memo(x-2)
    memo_list2[x] = ret
    return ret

#print("Fib: " + str(fib(fibval)) + "\nFib2: " + str(fib2(fibval)) + "\nFib memo: " + str(fib_memo(fibval)) + "\nFib2 memo: " + str(fib2_memo(fibval)))
"""print("fib:")
for i in range(fibval):
    print(str(fib(i))+" ", end=" ")
print("\nfib_memo:")
for i in range(fibval):
    print(str(fib_memo(i)) + " ", end=" ")
print("\nfib2:")
for i in range(fibval):
    print(str(fib2(i)) + " ", end=" ")

print("\nfib2_memo:")
for i in range(fibval):
    print(str(fib2_memo(i)) + " ", end=" ")
"""
print("fib(" + str(fibval) + "): " + str(fib2_memo(fibval)))

File: java/Fib.java
import java.util.Scanner;
import java.util.HashMap;

public class Fib{
    static HashMap<Integer, Integer> memo = new HashMap<>();

    public static void main(String[] args){
        Scanner s = new Scanner(System.in);
        int fibval = 0;

        System.out.print("Enter the value for the fib. sequence: ");
        fibval = s.nextInt();
        System.out.println("fibval: " + fibval);
        System.out.printf("Fib value for %d: %d\n", fibval, fib(fibval));
    }

    static int fib(int x){
        if(x < 0) return 0;
        if(x == 0 || x == 1){
            Fib.memo.put(x, x);
            return x;
        }
        if(Fib.memo.containsKey(x)){// && Fib.memo.size() > x){
            return Fib.memo.get(x);
        }
        int ret = fib(x-1) + fib(x-2);
        Fib.memo.put(x, ret);
        return ret;
    }
}
